src/dataflow/src/server.rs

```diff
@@ -218,19 +217,216 @@ pub fn serve(config: Config) -> Result<WorkerGuards<()>, String> {
+    /// compaction frontier.
+    fn compact(&mut self) {
+        let frontier = self.compaction_frontier.frontier();
+
+        if !frontier.is_empty() {
+            // Other folks are using these timestamp bindings. Lets keep everything
+            // >= what people are using
+            for (_, entries) in self.partitions.iter_mut() {
+                entries.retain(|(time, _)| frontier.less_equal(time));
+            }
+        }
+    }
+
+    /// Add a new mapping from `(partition, offset) -> timestamp`.
+    ///
+    /// Note that the `offset` and `timestamp` both have to be greater than the
+    /// largest previously bound offsets or timestamps for that partition.
+    fn add_binding(&mut self, partition: PartitionId, timestamp: Timestamp, offset: MzOffset) {
+        let entry = self.partitions.entry(partition).or_insert_with(Vec::new);
+        let (last_ts, last_offset) = entry.last().unwrap_or(&(0, MzOffset { offset: 0 }));
+        assert!(
+            offset >= *last_offset,
+            "offset should not go backwards, but {} < {}",
+            offset,
+            last_offset
+        );
+        assert!(
+            timestamp > *last_ts,
+            "timestamp should move forwards, but {} <= {}",
+            timestamp,
+            last_ts
+        );
+        entry.push((timestamp, offset));
+    }
+
+    /// Get the timestamp assignment for `(partition, offset)` if it is known.
+    ///
+    /// This function returns the timestamp and the maximum offset for which it is
+    /// valid.
+    fn get_binding(
+        &self,
+        partition: &PartitionId,
+        offset: MzOffset,
+    ) -> Option<(Timestamp, MzOffset)> {
+        if !self.partitions.contains_key(partition) {
+            return None;
+        }
+
+        let entries = self.partitions.get(partition).expect("known to exist");
+
+        for (ts, max_offset) in entries {
+            if offset <= *max_offset {
+                return Some((*ts, *max_offset));
+            }
+        }
+
+        return None;
+    }
+
+    /// Returns the maximum timestamp that is bound on all
+    /// of the partitions (e.g. the smallest timestamp that every
+    /// single partition knows about). We can guarantee that all
+    /// subsequent updates will be in advance of this timestamp.
+    fn get_upper(&self) -> Antichain<Timestamp> {
+        let mut upper = Antichain::new();
+
+        for (_, entries) in self.partitions.iter() {
+            if let Some((timestamp, _)) = entries.last() {
+                upper.insert(*timestamp);
+            }
+        }
+
+        upper
+    }
+
+    /// Returns the list of partitions this source knows about.
+    ///
+    /// TODO(rkhaitan): this function feels like a hack, both in the API of having
+    /// the source instances ask for the list of known partitions and in allocating
+    /// a vector to answer that question.
+    fn partitions(&self) -> Vec<PartitionId> {
```

Comment:
the comment says this but its pretty sucky that allocate a vec for this. the main point of this function is so that the source instances can ask "are there any new partitions I don't know about" and maybe should? Its entirely irrelevant for dbz byo sources because those always only have single partition topic.

Perhaps a better interface is one where the source instance supplies a iterator / list of partitions it already knows about then we return the set of partitions not in that iterator / list?

Reply:
Idk if this is Kafka specific (probably?) but if it isn't, I'd recommend we use the right words ("part" and "partition" vs "partitions" and "partitioning").

Reply:
You should be able to return an iterator I would think. You can use `self.partitions.keys` and have the return type either be explicit, or just `impl Iterator<Item = PartitionId>`. Or maybe you can't for some reason, but maybe that works?

Reply:
I'm sorry I don't quite understand

> Idk if this is Kafka specific (probably?) but if it isn't, I'd recommend we use the right words ("part" and "partition" vs "partitions" and "partitioning").

I don't see "partitioning" anywhere in the file

Reply:
Ignoring it is fine. It's a grammar thing where I want folks to say things like `PartId` rather than `PartitionId`. But the rest of the world has moved on I think.